packages/core/src/__tests__/unit/application-lifecycle.unit.ts

```diff
@@ -22,6 +22,46 @@ import {
 } from '../..';
 
 describe('Application life cycle', () => {
+  describe('state', () => {
+    it('updates application state', async () => {
+      const app = new Application();
+      expect(app.state).to.equal('created');
+      const start = app.start();
+      expect(app.state).to.equal('starting');
+      await start;
+      expect(app.state).to.equal('started');
+      const stop = app.stop();
+      expect(app.state).to.equal('stopping');
+      await stop;
+      expect(app.state).to.equal('stopped');
+    });
+
+    it('allows application.stop when it is created', async () => {
+      const app = new Application();
+      await app.stop(); // no-op
+      expect(app.state).to.equal('created');
```

Comment:
I find this confusing. How comes that after calling `app.stop`, the app can be either in `created` or `stopped` state? I would expect it to be always in `stopped` state. By allowing two different states, we are pushing more complexity on our users.

Maybe this is a hint that instead of having two states `created` and `stopped`, we should have a single state only, e.g. `not-running`?

Reply:
Good question. We may have a few choices:

1. Do not allow `stop`
2. Transition from `created` to `stopped`
3. Merge the state

Reply:
Originally, I wanted to propose to merge the two states and use `stopped` instead.

However, now that you have introduced `booting` & `booted` states, this is becoming even more confusing. Can we call `app.start()` before the application was booted? Or does it require the app to be in the state `booted`? Can we "un-boot" the application, i.e. go from `booted` back to `created`? I think that's not possible and not feasible to implement either.

So IIUC, we have the following states with the following allowed transitions:

1. `created` - can call `app.boot()` to go to `booting`
2. `booting` - no transition allowed, boot in progress, will go to `booted`
3. `booted` - app is booted but not started yet. Call `app.start()` to go to `starting` state
4. `starting` - no transition allowed, will go to `started`
5. `started` - can call `app.stop()` to go to `stopping` state
6. `stopping` - no transition allowed, will go to `stopped`
7. `stopped` - can call `app.start()` to go to `starting` state

If the list above is correct, then I am proposing to remove `stopped` state and move the app to `booted` state after it has stopped.

Thoughts?

/cc @strongloop/loopback-next